lmfdb/bianchi_modular_forms/test_bmf.py

```diff
@@ -113,3 +114,62 @@ def test_friends(self):
             # but the goal is to test that itself doesn't show in the friends list
             assert notitself not in L.get_data(as_text=True)
 
+    def check_compile_and_get_level(self, download_data):
+        """Simulates a user downloading the sage code, and then loading it into
+        a sage session. This requires the sage import at the top"""
+        aux_file = 'testing.sage'
+
+        with open(aux_file, 'w') as the_file:
+            the_file.write(download_data.get_data(as_text=True))
+
+        load(aux_file)  # if there's a problem with loading, this will make the test fail
+        os.remove(aux_file)
+        return NN
```

Comment:
This causes the one remaining pyflakes error, and I admit that the whole premise of this method might raise some eyebrows. Pyflakes complains because NN is seemingly not defined, but what it doesn't know is that the `load(aux_file)` defines it (it's the level of the BMF).

My intention with this method is to simulate a user downloading the sage code, and loading it in a sage session. This test will catch any errors in loading the sage code, which there are currently no checks for. It also returns the level ideal to the calling method, which then checks that it is equal to the ideal on the homepage.

Maybe there's another way of running this sage compilation check, one that in particular keeps pyflakes happy? I don't know what the LMFDB policy on declared exemptions to pyflakes/pylint etc are ...

Reply:
You should have no statements like `from foo import *`
Also, why are you writing to a file to load it? You should be doing something similar to what you did in magma, evaling the string, see for example: 
https://github.com/LMFDB/lmfdb/blob/master/lmfdb/classical_modular_forms/test_cmf.py#L575

Reply:
Aah cool, I didn't know about `exec`.

However, after trying it, there's an issue in that it's not finding the meaning of 'special' characters in sage, such as the `<` in `P.<x>` or `^`; a MWE is the single line in the sage console: `exec("print(x^2)")`. 

Do you know how I can get `exec` to work with this special sage notation?

(I've rewritten the sage code to not use `<` and `>`, but not using `^` makes things a bit more inconvenient, since it's entwined into the hecke poly.)

Reply:
I honestly recommend against all those shortcuts, as those are converted to valid python code with a preparser.

Here are various ways to create polynomials, without relying on the operator '^'

```
sage: R = PolynomialRing(ZZ, 'x')
sage: x = R.gen()
sage: f = 3*x**3 + 5
sage: R(list(f))
3*x^3 + 5
sage: list(f)
[5, 0, 0, 3]
sage: R("x^5 + 1")
x^5 + 1
```

Reply:
Ok thanks - I'll change the sage download code to output a list for the hecke polynomial, and then read that into the number field